src/bam/simulate_reads.rs

```diff
@@ -0,0 +1,181 @@
+    let mut artificial_record = bam::record::Record::new();
+    if let Ok(mate_cigar) = record.aux(b"MC") {
+        artificial_record.push_aux(b"MC", mate_cigar)?;
+    }
+    artificial_record.set(
+        record.qname(),
+        Some(&record.cigar()),
+        artificial_seq,
+        record.qual(),
+    );
+    artificial_record.set_pos(record.pos() - offset);
+    artificial_record.set_tid(0);
+    artificial_record.set_mtid(0);
+    artificial_record.set_mpos(record.mpos() - offset);
+    artificial_record.set_flags(record.flags());
+    artificial_record.set_insert_size(record.insert_size());
+    artificial_record.set_mapq(record.mapq());
+    Ok(artificial_record)
+}
+
+fn build_sequence(
+    altered_bases: &mut HashMap<usize, HashMap<u8, u8>>,
+    record: &bam::Record,
+    offset: usize,
+    rng: &mut ThreadRng,
+    alphabet: &[u8],
+) -> Result<Vec<u8>> {
+    let mut artificial_seq = Vec::new();
+    let record_seq = record.seq().as_bytes();
+    let mut record_pos = 0;
+    let mut ref_pos = record.pos() as usize - offset;
+    //Create random seq for leading softclips
+    for cigar in record.cigar_cached().unwrap().iter() {
+        match cigar.char() {
+            'S' => {
+                add_random_bases(cigar.len() as u64, &mut artificial_seq, rng, alphabet)?;
+                record_pos += cigar.len() as usize;
+            }
+            'M' | 'X' | '=' => {
+                (0..cigar.len()).for_each(|_| {
+                    let mut reduced_alphabet = alphabet.to_vec();
+                    let base_mappings = altered_bases.get(&ref_pos).unwrap().clone();
+                    let altered_base = *altered_bases
+                        .get_mut(&ref_pos)
+                        .unwrap()
+                        .entry(*record_seq.get(record_pos).unwrap())
+                        .or_insert_with(|| {
+                            reduced_alphabet.retain(|x| !base_mappings.values().any(|y| x == y));
+                            let idx = rng.gen_range(0, reduced_alphabet.len());
+                            reduced_alphabet[idx]
+                        });
+                    artificial_seq.push(altered_base);
+                    ref_pos += 1;
+                    record_pos += 1;
+                });
+                // Add reference bases except for mismatches
+            }
+            'I' => {
+                add_random_bases(cigar.len() as u64, &mut artificial_seq, rng, alphabet)?;
+                record_pos += cigar.len() as usize;
+            }
+            'D' | 'N' => {
+                ref_pos += cigar.len() as usize;
+            }
+            _ => {}
+        }
+    }
+
+    Ok(artificial_seq)
+}
+
+fn add_random_bases(
+    length: u64,
+    seq: &mut Vec<u8>,
+    rng: &mut ThreadRng,
+    alphabet: &[u8],
+) -> Result<()> {
+    let alphabet_size = alphabet.len();
+    (0..length).for_each(|_| {
+        let idx = rng.gen_range(0, alphabet_size);
+        seq.push(alphabet[idx])
```

Comment:
I think here you can use [choose](https://docs.rs/rand/0.8.4/rand/seq/trait.SliceRandom.html#tymethod.choose), should be more ergonomic.